src/net_processing.cpp

```diff
@@ -1252,6 +1253,33 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
            (GetBlockProofEquivalentTime(*pindexBestHeader, *pindex, *pindexBestHeader, m_chainparams.GetConsensus()) < STALE_RELAY_AGE_LIMIT);
 }
 
+bool PeerManagerImpl::FetchBlock(const NodeId nodeid, const CBlockIndex* pindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    PeerRef peer = GetPeerRef(nodeid);
+    if (peer == nullptr) return false;
+    uint32_t nFetchFlags = 0;
+    CNodeState* state = State(nodeid);
+    if (state == nullptr) {
+        return false;
+    }
+    const int node_sync_height = state->pindexBestKnownBlock ? state->pindexBestKnownBlock->nHeight : -1;
+    if (pindex->nHeight > node_sync_height) {
+        return false;
+    }
+    if (state->fHaveWitness) {
+        nFetchFlags |= MSG_WITNESS_FLAG;
+    }
+    std::vector<CInv> vInv(1);
+    vInv[0] = CInv(MSG_BLOCK | nFetchFlags, pindex->GetBlockHash());
+    bool success = m_connman.ForNode(nodeid, [this, vInv](CNode* pnode) {
+        const CNetMsgMaker msgMaker(pnode->GetCommonVersion());
+        this->m_connman.PushMessage(pnode, msgMaker.Make(NetMsgType::GETDATA, vInv));
+        return true;
+    });
+    MarkBlockAsInFlight(nodeid, pindex->GetBlockHash(), pindex);
```

Comment:
Perhaps move this above the `ForNode()` call and don't send the `getdata` if it returns false, since that indicates that the block is already in flight. We don't want a faulty RPC client causing us to spam a peer with multiple requests for the same block.

Reply:
Not sure if I got this right now...